Resource packs: support dynamic input tex

iwubcode opened the pull request:
This builds on #8318 by adding support for Resource Packs.  It makes Resource Packs more generic so they deal with a file that is a particular "ResourceType" (currently just texture or dynamic input).

It also does two other things:

* In master, if you moved priority up/down of a pack.  It would uninstall the pack first but never reinstall it.  This seemed like an oversight and has been fixed
* A pack can be installed or uninstalled (or priority changed) when the game is running.  Texture only packs reload the textures and the cache is invalidated.  For dynamic texture packs, the textures are regenerated (and textures are reloaded / cache is invalidated).

lioncash commented on Source/Core/InputCommon/ImageOperations.cpp:
```diff
@@ -0,0 +1,381 @@
+  ImagePixelData image;
+  std::vector<u8> data;
+  if (!Common::LoadPNG(buffer, &data, &image.width, &image.height))
+    return {};
```
`return std::nullopt;`

The nullopt constructor and default constructor invocations don't always do the [same thing](https://godbolt.org/z/UHP5B9)

iwubcode replied:
Very interesting, I wasn't aware of that.  Thank you!

lioncash commented on Source/Core/InputCommon/ImageOperations.cpp:
```diff
@@ -0,0 +1,381 @@
+
+  std::vector<Rect> regions;
+  std::vector<bool> seen_pixels;
+  seen_pixels.resize(src.width * src.height);
```
This can be collapsed into the constructor:

```suggestion
  std::vector<bool> seen_pixels(src.width * src.height);
```

or, this could be moved entirely into CalculateNeighboringPixels (see next review comment). It could also be dropped entirely from the interface, given it's not actually used, aside from being stored into in CalculateNeighboringPixels.

iwubcode replied:
Unless I'm missing something from your suggestion, I don't think `seen_pixels` can move into `CalculateNeighboringPixels` because it is built up from multiple for loop passes.  The other vector (as you mention below) could be moved to be a return value instead though.

lioncash replied:
Ah OK, I must have missed the read from the vector in the function when doing my initial review. Disregard the latter half of my comment.

lioncash commented on Source/Core/InputCommon/ImageOperations.cpp:
```diff
@@ -0,0 +1,381 @@
+  return src.pixels[x + y * src.width];
+}
+
+void CalculateNeighboringPixels(const ImagePixelData& src, std::vector<Point>& neighboring_pixels,
```
These vectors can be returned as a pair, which allows them to be unpacked with structured bindings.

lioncash commented on Source/Core/InputCommon/ImageOperations.cpp:
```diff
@@ -0,0 +1,381 @@
+
+std::vector<Rect> FindRectRegionsOfNeighboringPixels(const ImagePixelData& src)
+{
+  static const Pixel mask_color{0, 0, 0, 255};
```
```suggestion
  static constexpr Pixel mask_color{0, 0, 0, 255};
```

lioncash commented on Source/Core/InputCommon/ImageOperations.cpp:
```diff
@@ -0,0 +1,381 @@
+
+ImagePixelData Resize(ResizeMode mode, const ImagePixelData& src, u32 new_width, u32 new_height)
+{
+  ImagePixelData result;
```
This and the following three lines can be collapsed into:

```suggestion
  ImagePixelData result(new_width, new_height);
```

iwubcode commented:
Thank you @lioncash for your feedback.  Lots of good stuff.

iwubcode commented on Source/Core/InputCommon/DynamicInputTextureConfiguration.h:
```diff
@@ -0,0 +1,48 @@
+  void GenerateTextures(const IniFile::Section* sec);
+
+private:
+  struct DynamicInputTextureData
```
I like this idea but the vector's type but VisualStudio complains about this change.  I did a similar forward declaration of `DynamicInputTextureConfiguration` in `DynamicInputTextureManager`'s header but with this forward declaration it complains there about not having the size of `DynamicInputTextureConfiguration`.  I'm not too familiar with this style of forward declaration, I mostly forward declare pointer types, so I might be missing something.

lioncash replied:
Can you provide both the header and source file with the changes you made that caused the compilation error?

iwubcode replied:
@lioncash - please see the last commit on #8318

iwubcode replied:
To give an example:

```
// a.h
struct B;
struct A
{
 A();
 ~A();
 B b;
};
void do_something(const& A a);

// a.cpp
#include "b.h"
A::A() = default;
A::~A() = default;
void do_something(const& A a)
{
  // ...
}

// b.h
struct B
{
  B();
  ~B();
  struct Imp;
  imp imp;
};

// B.cpp
struct B::Imp
{
  int a = 3;
};
B::B() = default;
B::~B() = default;
```

(didn't actually compile this but it portrays the basic idea) We can't compile "A" because "B" entire size isn't known from within A's cpp file

iwubcode replied:
I reverted the change (since it wasn't building as is), it only saves a map include.  Let me know though if I'm missing something.

BhaaLseN commented on Source/Core/UICommon/ResourcePack/ResourcePack.cpp:
```diff
@@ -19,6 +21,7 @@
 namespace ResourcePack
 {
 constexpr char TEXTURE_PATH[] = "Load/Textures/";
+constexpr char DYNAMIC_INPUT_PATH[] = "Load/DynamicInputTextures";
```
`TEXTURE_PATH` has a trailing slash, while this one does not. Does it make a difference to use either for consistency? I for one prefer folder paths to end in a slash so I don't have to worry about it elsewhere.

BhaaLseN commented on Source/Core/InputCommon/InputCommon.vcxproj:
```diff
@@ -35,6 +35,16 @@
   <PropertyGroup Label="UserMacros" />
+  <ItemDefinitionGroup Condition="'$(Configuration)|$(Platform)'=='Release|x64'">
+    <ClCompile>
+      <AdditionalIncludeDirectories>C:\dev\opencv\build\include;%(AdditionalIncludeDirectories)</AdditionalIncludeDirectories>
```
I don't think we should be committing a local path like `C:\dev\opencv` here. Also, are those two `ItemDefinitionGroup` now obsolete that you removed the OpenCV dependency?

iwubcode replied:
Yes this whole block can be removed, thanks

iwubcode commented on Source/Core/InputCommon/DynamicInputTextureConfiguration.cpp:
```diff
@@ -0,0 +1,251 @@
+  File::OpenFStream(json_stream, json_file, std::ios_base::in);
+  if (!json_stream.is_open())
+  {
+    ERROR_LOG(PAD, "Failed to load dynamic input json file '%s'", json_file.c_str());
```
Do you have a suggestion for a logger?  It was a tossup between PAD and one of the graphics ones.  Since this is for inputs, I picked PAD

BhaaLseN replied:
Very good point. I'd rather use `VIDEO` (or not?) because it changes something graphics related (ie. the result is shown on screen; or _not_ when something is wrong). Guessing Screenshots/FrameDumps go to `VIDEO` for that reason.
Or we define a new one; perhaps specific to resource packs, perhaps generic enough to handle many different customization points?
I can live with either, but feel like `PAD` is the last one of them that I'd look at for a _texture_ related message.

iwubcode replied:
Yeah, I originally thought about making my own category but for such a minor feature, I didn't know if it was worth it.  I don't really mind one way or other so I'll move it to `VIDEO`.

iwubcode commented on Source/Core/InputCommon/DynamicInputTextureConfiguration.cpp:
```diff
@@ -0,0 +1,251 @@
+    return;
+  }
+  const std::string json_str((std::istreambuf_iterator<char>(json_stream)),
+                             (std::istreambuf_iterator<char>()));
```
Coming back to this, this is indeed cryptic and quite slow.  I will change it

iwubcode commented on Source/Core/InputCommon/ImageOperations.cpp:
```diff
@@ -0,0 +1,380 @@
+
+#ifdef _MSC_VER
+#pragma warning(push)
+#pragma warning(disable : 4611)
```
Done, thanks!

iwubcode commented:
A big thank you to @BhaaLseN  for reviewing.  I updated the code based on the feedback.

I also pulled in @spycrab 's original ResourcePack document which used to live outside the Dolphin repo (here:  https://gist.github.com/spycrab/9d05056755d8d7908bdb871a99d050bf) and then I updated it with the new dynamic input information (need to figure out how to link to another md file).

BhaaLseN commented on CustomFileFormats/ResourcePacks.md:
```diff
@@ -0,0 +1,114 @@
+# Dolphin Resource Pack Specification v3
```
Any reason you didnt just update `docs/ResourcerPacks.md` instead?

iwubcode replied:
I saw `docs` but missed that `ResourcePacks.md` was in there.  Thanks!

BhaaLseN commented on CustomFileFormats/DynamicInputTextures.md:
```diff
@@ -0,0 +1,136 @@
+# Dolphin Dynamic Input Textures Specification (v1)
```
Just like `ResourcePacks.md`, this should probably go in the `docs/` folder.

iwubcode replied:
Will do, thank you!

BhaaLseN commented on CustomFileFormats/DynamicInputTextures.md:
```diff
@@ -0,0 +1,136 @@
+
+*colorkey_image* - an image with colored boxes, will be used as a base image, so any non-input related sections can be upscaled/redrawn if desired
+
+*fallback_image* - after looking at the colorkey_image, if the color isn't mapped to a specific button, it will use this image as a reference for what to put in the color. This should be the same size/quality as the colorkey_image but without the colored boxes
```
I personally prefer identifiers to be monospaced, and at a glance it seems `ResourcePacks.md` also does this.
Would you mind updating them here?
```suggestion
*fallback_image* - after looking at the ``colorkey_image``, if the color isn't mapped to a specific button, it will use this image as a reference for what to put in the color. This should be the same size/quality as the ``colorkey_image`` but without the colored boxes
```

iwubcode replied:
Did you mean instead of being italicized?  I had made one italicized to visually differentiate them from the table which is monospaced.  I can try monospacing everything though.  I wanted the descriptions in the table but they are much longer than the ones in ResourcePacks

BhaaLseN replied:
I mainly mean use of previously (emphasised or not) identifiers to set them apart from plain-text; whether thats italicized or monospaced or whatever is up to you (as long as it is consistent)

iwubcode replied:
Ok updated.  Please let me know if this isn't what you had in mind.

BhaaLseN commented on Source/Core/UICommon/ResourcePack/ResourcePack.cpp:
```diff
@@ -78,26 +86,49 @@ ResourcePack::ResourcePack(const std::string& path) : m_path(path)
 
   unzGoToFirstFile(file);
 
+  std::array<ResourceType, 2> resource_types = {ResourceType::TEXTURE, ResourceType::DYNAMIC_INPUT};
```
Can (or should) this be `static`, or perhaps even `constexpr`?

BhaaLseN commented on Source/Core/UICommon/ResourcePack/ResourcePack.cpp:
```diff
@@ -78,26 +86,49 @@ ResourcePack::ResourcePack(const std::string& path) : m_path(path)
+
+      if (filename.compare(0, zip_directory_root.size(), zip_directory_root) != 0 ||
+          file_info.uncompressed_size == 0)
+        continue;
```
Since this `if` condition spans multiple lines, curlies should be used for the body.

BhaaLseN commented on docs/DynamicInputTextures.md:
```diff
@@ -0,0 +1,139 @@
+*fallback_image* - after looking at the ``colorkey_image``, if the color isn't mapped to a specific button, it will use this image as a reference for what to put in the color. This should be the same size/quality as the ``colorkey_image`` but without the colored boxes.
+
+*emulated_controls* - a series of emulated buttons that map to a given color rectangle.
+
```
Did you intentionally leave _two_ blank lines here between explainations, vs. one between every other?

iwubcode replied:
Whoops, no that wasn't intentional

RinMaru commented:
is it possible to use better quality textures?

Rumi-Larry commented:
What happened with this pr?

iwubcode commented:
While this is perfectly mergeable as is.  I have recently had an idea that might affect this PR.  If in the future this PR is still relevant, I'll re-open.
